i6_models/parts/conformer/frontend.py

```diff
@@ -0,0 +1,279 @@
+        model_cfg.check_valid()
+
+        conv_padding = self._get_padding(model_cfg.conv_kernel_size)
+        pool_padding = self._get_padding(model_cfg.pool_kernel_size)
+
+        self.conv1 = nn.Conv2d(
+            in_channels=model_cfg.features,
+            out_channels=model_cfg.channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.conv2 = nn.Conv2d(
+            in_channels=model_cfg.channels,
+            out_channels=model_cfg.channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.pool = nn.MaxPool2d(
+            kernel_size=model_cfg.pool_kernel_size,
+            stride=model_cfg.pool_strides,
+            padding=pool_padding,
+        )
+        self.conv3 = nn.Conv2d(
+            in_channels=model_cfg.channels,
+            out_channels=model_cfg.channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.conv4 = nn.Conv2d(
+            in_channels=model_cfg.channels,
+            out_channels=model_cfg.channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.activation = model_cfg.activation
+
+    def forward(self, tensor: torch.Tensor) -> torch.Tensor:
+        """
+        T might be reduced to T' depending on ...
+
+        :param tensor: input tensor of shape [B,T,F]
+        :return: torch.Tensor of shape [B,T',F]
+        """
+        # conv layers expect shape [B,F,T] so we have to transpose here
+        tensor = torch.transpose(tensor, 1, 2)  # [B,F,T]
+        tensor = tensor[:, :, :, None]  # [B,F,T,C]
+        tensor = self.conv1(tensor)
+        tensor = torch.transpose(tensor, 1, 2)  # transpose back to [B,T,F,C]
+        tensor = torch.squeeze(tensor)  # [B,T,F]
+
+        tensor = self.activation(tensor)
+        tensor = self.pool(tensor)
+
+        tensor = torch.transpose(tensor, 1, 2)  # [B,F,T]
+        tensor = tensor[:, :, :, None]  # [B,F,T,C]
+        tensor = self.conv2(tensor)
+        tensor = torch.transpose(tensor, 1, 2)  # transpose back to [B,T,F,C]
+        tensor = torch.squeeze(tensor)  # [B,T,F]
+
+        tensor = self.activation(tensor)
+
+        tensor = torch.transpose(tensor, 1, 2)  # [B,F,T]
+        tensor = tensor[:, :, :, None]  # [B,F,T,C]
+        tensor = self.conv3(tensor)
+        tensor = torch.transpose(tensor, 1, 2)  # transpose back to [B,T,F,C]
+        tensor = torch.squeeze(tensor)  # [B,T,F]
+
+        tensor = self.activation(tensor)
+
+        tensor = torch.transpose(tensor, 1, 2)  # [B,F,T]
+        tensor = tensor[:, :, :, None]  # [B,F,T,C]
+        tensor = self.conv4(tensor)
+        tensor = torch.transpose(tensor, 1, 2)  # transpose back to [B,T,F,C]
+        tensor = torch.squeeze(tensor)  # [B,T,F]
+
+        return tensor
+
+
+def _get_padding(input_size: Union[int, Tuple[int, ...]]) -> int:
```

Comment:
I wonder a bit why you use the padding at all. Is this intentional? I think e.g. in ESPnet, they don't add any padding in the frontend. I think in most other implementations, they would not add padding. But I think in our TF-based setups, we add padding, but probably only because people never really thought about it and it was the default for our `ConvLayer`. Maybe you also want to have it configurable?

Further, maybe you might want padding only in the time dimension, but not in the other dimension?

Reply:
> Uses explicit padding for ONNX exportability, see: https://github.com/pytorch/pytorch/issues/68880

from Nick, I think the explicit padding is then here also necessary

Reply:
No, what you link is about `padding="same"` vs `padding=<some non-zero int>`. Or `padding="valid"` vs `padding=0`.

When I say, why padding at all, I mean, why not `padding=0`?

Reply:
the idea was so that the time dimension does not change. without padding it seems that we loose a few frames.. I made this more configurable, so that the user has more influence on this